crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2372,11 +2372,49 @@ impl<'db> TypeInferenceBuilder<'db> {
             ctx: _,
         } = subscript;
 
-        self.infer_expression(slice);
-        self.infer_expression(value);
+        let value_ty = self.infer_expression(value);
+        let slice_ty = self.infer_expression(slice);
 
-        // TODO actual subscript support
-        Type::Unknown
+        match (value_ty, slice_ty) {
+            // TODO handle variable-length tuples
+            (Type::Tuple(ty), Type::IntLiteral(index)) => {
+                if let Some(ty) = ty.elements(self.db).get(index as usize) {
+                    *ty
+                } else {
+                    Type::Unknown
+                }
+            }
+            (Type::Tuple(ty), Type::BooleanLiteral(bool)) => {
+                if let Some(ty) = ty.elements(self.db).get(usize::from(bool)) {
+                    *ty
+                } else {
+                    Type::Unknown
+                }
+            }
+            (Type::StringLiteral(literal), Type::IntLiteral(index)) => {
+                if let Some(ch) = literal.value(self.db).chars().nth(index as usize) {
+                    Type::StringLiteral(StringLiteralType::new(
+                        self.db,
+                        ch.to_string().into_boxed_str(),
+                    ))
+                } else {
+                    Type::Unknown
+                }
+            }
+            (Type::StringLiteral(literal), Type::BooleanLiteral(index)) => {
+                if let Some(ch) = literal.value(self.db).chars().nth(usize::from(index)) {
+                    Type::StringLiteral(StringLiteralType::new(
+                        self.db,
+                        ch.to_string().into_boxed_str(),
+                    ))
+                } else {
+                    Type::Unknown
+                }
+            }
+            // TODO some slices on strings should resolve to type errors rather than literals
+            (Type::StringLiteral(_), _) => Type::LiteralString,
```

Comment:
What's the general philosophy around when we should error in a case like this?

Reply:
There's a few things to note here.

### Issues with `LiteralString`

According to a strict reading of [PEP 675](https://peps.python.org/pep-0675/#type-inference), I don't think this is a legal place where we can infer `LiteralString`, since the PEP states:

> As per the [Rationale](https://peps.python.org/pep-0675/#rationale), we also infer `LiteralString` in the following cases:
> 
> [...]
>
> - Literal-preserving methods: In [Appendix C](https://peps.python.org/pep-0675/#pep-675-appendix-c), we have provided an exhaustive list of `str` methods that preserve the `LiteralString` type.

`str.__getitem__` is not listed in Appendix C of the PEP, so I suppose that means that we're not allowed to infer `LiteralString` for `x`, `y` or `z` in a snippet like this:

```py
def foo(a: Literal["foo"], b: LiteralString, c: int):
    x = a[c]
    y = b[42]
    z = b[c]
```

I'm not sure what the status of Appendix C is, though... it doesn't seem to have been copied into the [relevant section](https://typing.readthedocs.io/en/latest/spec/literal.html#valid-locations-for-literalstring) of the typing spec, so perhaps it isn't canonical anymore? Or maybe that was just an omission when writing the spec? There's also the question of typeshed's stubs: typeshed doesn't include a `LiteralString` overload in its stub for `str.__getitem__`, although it includes `LiteralString` overloads for many of its other `str` methods. I'm not sure if there's a reason for that or not -- even if not, it may be desirable to be in sync with typeshed here.

For now I think I'd just delete this branch -- continuing to fallback to `Unknown` is ~fine for now.

### Emitting diagnostics

In general, if we can't infer that an operation creates a `Literal` type (whether that's `LiteralString` or `StringLiteral`), we should fall back to typeshed's stubs. Here that means falling back to `builtins.str.__getitem__` -- you'd do something like `value_ty.to_meta_type(db).member(db, "__getitem__").call(&[value_ty, slice_ty])` to figure out what the type of the subscript expression is. Except there are some complications ;) Take a look at `Type::iterate` for a more complete example of how we lookup dunder methods elsewhere, try to call them, and emit a diagnostic if the dunder method eithere doesn't exist or isn't callable. https://github.com/astral-sh/ruff/blob/5118166d21784e7c78e38ea42919ba50bb2a5142/crates/red_knot_python_semantic/src/types.rs#L629-L684

Reply:
I don't agree with the pedantic reading of PEP 675 (or typeshed) as a rationale for removing this branch. Many useful things are simply overlooked. If a more precise type inference is clearly correct in terms of the runtime semantics, and there's nothing in the typing spec or conformance suite explicitly prohibiting it, then I think we should feel totally free to go ahead and do it.

Reply:
On second thought, given the security rationale for `LiteralString`, there _may_ be an argument that untrusted-user-supplied indices could allow turning a "safe" `LiteralString` into an "unsafe" one via careful cropping. I have mixed feelings about this, because there's also type-system value to `LiteralString` (we know it is really a built-in `str`, not a subclass), and in type-system terms, the inference of `LiteralString` is correct. And the security issue seems fairly contrived. But that _might_ be why inference of `LiteralString` on `__getitem__` is not in the PEP?

Reply:
> It is clearly correct wrt the runtime semantics to infer `LiteralString` as the result of indexing into a `LiteralString` or `StringLiteral`.

I agree, but the fact that PEP-675 is so clear that the list of string methods in Appendix C is valid is meant to be "exhaustive" made me worried that there was something I was misunderstanding :-) and until recently, PEP-675 _was_ the spec -- it's not clear to me whether Appendix C was deliberately kept out of the typing spec or not.

This branch was anyway incorrect -- `StringLiteral` cannot be indexed into with objects of arbitraty types, which is what this branch allowed before Charlie removed it.

Reply:
After further thought, I don't really buy the possible security argument I mentioned above for not inferring `LiteralString`: you can easily construct many other cases where the contents of a `LiteralString`, though always originating from a literal string in the source code, could be _influenced_ in some way by untrusted user input (e.g. determining whether some concatenation does or doesn't occur.)

But that said, I don't think this is critical, and it is probably best to just rely on typeshed for this. So we could at some point consider proposing a change to typeshed with a `LiteralString` overload for `str.__getitem__`, but I don't think it's worth an extra special case in red-knot to diverge from typeshed here.

So yeah, I would say let's just add the branch that uses the return type from `__getitem__`, as Alex suggested.